### DISCUSSION TITLE
very slow progress running MET V10.0 Series-Analysis

### DISCUSSION THREAD

**Discussion by jwraby** [dtcenter/METplus#1255] [Statistical Computation] [answered]
I ran Series-Analysis using separate -fcst and -obs input which successfully completed in about 3mins on an HPC. Block size was 100000. When I use the -both option to specify the input of the same fcst and obs data from the NetCDF output of Grid-Stat, it runs very slowly using the same block size. I increased the block size to 1000000, but no change. Has similar behavior been reported by others? What is a likely cause of this slow execution?
Thanks.

**Comment by CPKalb** [comment]
Hi:

Could you send over the command lines that are being used to run the separate -fcst and -obs case and the -both case?  Also, do you have log files for these two runs that we could take a look at?  The log files may help us understand what is going on.

Thanks,
Christina

**Reply by jwraby** [reply]
Hi Christina - Sorry for my delayed response. I have a text file containing the two commands I used for both cases and I also have the log files generated for both cases. I attached the 3 files to this posting
[series_analysis_v10.0_test_commands.txt](https://github.com/dtcenter/METplus/files/7547517/series_analysis_v10.0_test_commands.txt)
[run_series_analysis_m3o3_4day_TMP_winter_cases_both.log](https://github.com/dtcenter/METplus/files/7547520/run_series_analysis_m3o3_4day_TMP_winter_cases_both.log)
[run_series_analysis_m3o3_4day_TMP_winter_cases.log](https://github.com/dtcenter/METplus/files/7547521/run_series_analysis_m3o3_4day_TMP_winter_cases.log)

.

**Reply by CPKalb** [reply]
Thanks.  To clarify, is one run using netCDF and the other using grib data?  I want to make sure I understand exactly how these are run.

Christina

**Comment by jwraby** [comment]
CLASSIFICATION: UNCLASSIFIED

 

Yes. The –both run is using netCDF. The –fcst data is grib (from the UPP) and the –obs data is .grb2

R/

John

**Reply by CPKalb** [reply]
Hi John:

Thanks.  Would you be able to send us a copy of your data and the config files for the runs.  Instructions to send us data can be found here: 
https://github.com/dtcenter/METplus/discussions/954

Thanks,
Christina

**Reply by JohnHalleyGotway** [reply]
@jwraby I see one subtle difference in the log files you sent.
From the "both" log:
```
DEBUG 2: Computing statistics using a block size of 1000000, requiring 1 pass(es) through the 204 x 204 grid.
```
From the "fcst" and "obs" log:
```
DEBUG 2: Computing statistics using a block size of 100000, requiring 1 pass(es) through the 204 x 204 grid.
```
Notice the one extra 0 in the first one. I suspect that Series-Analysis is allocating memory for several arrays up front based on the specified block size. And allocating arrays 10 times longer takes a lot longer. Please try re-running the "both" case with:
```
block_size = 204 * 204;
```
Does that produce improved runtimes? If so, we could consider updating Series-Analysis to check if block_size > Nx * Ny, and if so, automatically reduce it to Nx * Ny.

**Comment by jwraby** [comment]
CLASSIFICATION: UNCLASSIFIED

 

Hi John – 

 

That difference in block size was not always the case. My first run with –both used the same size as the run with –fcst/-obs (100000). I incrementally increased the block size for subsequent runs to see if that would fix the problem and nothing changed. I stopped the testing after the last run which used 1000000 because I saw no improvement. 

 

I will try your suggested change by editing my config file to specify a block size of 204 * 204 vice 1000000.

 

R/

John

**Reply by JohnHalleyGotway** [reply]
@jwraby for what it's worth, I did some additional testing and found behavior similar to what you describe.

I took GRIB1 data from a call to Grid-Stat, along with the NetCDF matched pair output from Grid-Stat, and used that data to call Series-Analysis 3 different ways:
1. Using -fcst and -obs to pass the same input GRIB files 100 times took about **18 seconds** on my machine.
2. Using -both to pass the same NetCDF matched pairs file 100 times took **93 seconds** on my machine.
3. Using -both to pass the same input GRIB file 100 times took about **18 seconds** on my machine.

These times are consistent with the behavior you describe. However, it isn't the difference between -fcst/-obs and -both that are the culprit. Reading data from GRIB input files appears to be much faster than reading from NetCDF input files.

I don't have a good explanation for why that's the case, but it certainly looks like it is!

I've attached my test data and script for reference:
[example.tar.gz](https://github.com/dtcenter/METplus/files/7558275/example.tar.gz)

**Comment by jwraby** [comment]
CLASSIFICATION: UNCLASSIFIED

 

Thanks for running this testing scenario. Very interesting! I just kicked off a run on the Navy HPC using your suggested value of 204*204 for block size, but the screen dialog halted a few minutes after reading the config file and then it read the first series entry and halted again. See below:

 

DEBUG 1: Default Config File: /usr/local/share/met/config/SeriesAnalysisConfig_default

DEBUG 1: User Config File: /p/home/jraby/MET/MET_SeriesAnalysis/SeriesAnalysisConfig_m3o3_TMP_tng

GSL_RNG_TYPE=mt19937

GSL_RNG_SEED=1

DEBUG 1: Length of configuration "fcst.field" = 1

DEBUG 1: Length of configuration "obs.field"  = 1

DEBUG 1: Length of forecast file list         = 100

DEBUG 1: Length of observation file list      = 100

DEBUG 1: Series defined by the forecast file list of length 100.

DEBUG 2: Processing series entry 1 of 100: FCST_TMP_Z2_FULL/Z2 versus OBS_TMP_Z2_FULL/Z2

DEBUG 2: Found data for FCST_TMP_Z2_FULL/Z2 in file: /p/home/jraby/MET/MET_GridStat/results_MSAregion_large-scale_Dawson/20161111_m1o3/grid_stat_WREN_assess_d03_030000L_20161111_120000V_pairs.nc

DEBUG 2: Processing masking regions.

DEBUG 2: Computing statistics using a block size of 41616, requiring 1 pass(es) through the 204 x 204 grid.

DEBUG 2: Processing data pass number 1 of 1 for grid points 1 to 41616.

DEBUG 1: Reading stat column descriptions: /usr/local/share/met/table_files/stat_column_description.txt

DEBUG 2: Processing series entry 2 of 100: FCST_TMP_Z2_FULL/Z2 versus OBS_TMP_Z2_FULL/Z2

 

The delay between processing entry 1 and entry 2 was like 5 mins or so.

 

This behavior is identical to the previous runs using –both. Appears that the 204*204 didn’t make any difference, but if the issue is in reading GRIB vs NetCDF then maybe the block size is not the culprit as you mentioned. 

 

R/

John

**Reply by JohnHalleyGotway** [reply]
5 minutes!? Yikes! 204 x 204 certainly is NOT a lot of grid points. Are you able to send us that NetCDF matched pairs file for additional investigation?
```
/p/home/jraby/MET/MET_GridStat/results_MSAregion_large-scale_Dawson/20161111_m1o3/grid_stat_WREN_assess_d03_030000L_20161111_120000V_pairs.nc
```
I'm curious if the read time is affected by the number of variables in the NetCDF file. Here's instructions for [How To Send Us Data](https://github.com/dtcenter/METplus/discussions/954).

**Comment by jwraby** [comment, marked as answer]
CLASSIFICATION: UNCLASSIFIED

 

I’m downloading that file now (519 MB) and will investigate my ability to use ftp which has been blocked for Army users in the past. I will investigate this, but I may have to upload the file using the SAFE file exchange system. If I can’t use the ftp, is it OK to use your email address to send a download notice/link?

**Reply by JohnHalleyGotway** [reply]
@jwraby, that's just fine. Feel free to send the SAFE file exchange directly to my personal email address.
I'm guessing there are ways we could make MET's reading of those NetCDF files more efficient.
However, we might also be able to modify your Grid-Stat configuration file to write much less data to those NetCDF matched pairs file. And I'm guessing that the smaller they are, the more quickly MET will be able to read them.

So please also send along your Grid-Stat Configuration file for me to take a look.

**Reply by JohnHalleyGotway** [reply]
@jwraby thanks for sending the sample data file. I see that it contains a WHOPPING 2929 data variables:
```
ncdump -h grid_stat_WREN_assess_d03_030000L_20161111_120000V_pairs.nc | grep "float "  | wc -l
    2929
```
And these are all for only 9 variables being evaluated in Grid-Stat:
```
ncdump -h grid_stat_WREN_assess_d03_030000L_20161111_120000V_pairs.nc | grep "float " | grep FCST | cut -d'_' -f2-4 | cut -d'(' -f1 | sort -u
DPT_Z2_FULL
GUST_Z0_FULL
SPFH_Z2_FULL
TCDC_L0_FULL
TMP_Z2_FULL
UGRD_Z10_FULL
VGRD_Z10_FULL
VIS_Z0_FULL
WIND_Z10_FULL
```
Only 20 of the variables do NOT contain NBRHD, thresholded output, or difference fields:
```
ncdump -h grid_stat_WREN_assess_d03_030000L_20161111_120000V_pairs.nc | grep "float "  | egrep -v "NBRHD|_ge|_le|DIFF_" | wc -l
      20
```
Trimming that file down from 2929 variables to the 20 listed above reduces the file size from 495M to 3.2M.
Thought I'd include the nccopy command because it was a little crazy!
```
nccopy -V `ncdump -h grid_stat_WREN_assess_d03_030000L_20161111_120000V_pairs.nc | grep "float "  | egrep -v "NBRHD|_ge|_le|DIFF"  | cut -d' ' -f2 | cut -d'(' -f1 | sed -r 's/$/,/g' | tr -d '\n' | sed -r 's/,$//g'` grid_stat_WREN_assess_d03_030000L_20161111_120000V_pairs.nc subset.nc
```
I ran a similar timing test, passing the same input file to Series-Analysis 100 times.

Processing that subset.nc file 100 times in Series-Analysis takes on 9 seconds... which is 1/2 the time that the GRIB input took!

And I stopped waiting for the full input file to run of Series-Analysis to complete. Suffice it to say, it takes a ridiculously long amount of time.

I'd recommend that you think carefully about what series analysis runs you'd like to perform. And only include the output variables necessary to complete them in the Grid-Stat matched pairs output file.

Take a look at the Grid-Stat [configuration options](https://met.readthedocs.io/en/latest/Users_Guide/grid-stat.html#grid-stat-configuration-file)... specifically the "nc_pairs_flag" option. Consider setting it like this.
```
nc_pairs_flag = {
   latlon       = FALSE;
   raw          = TRUE;
   diff         = FALSE;
   climo        = FALSE;
   climo_cdp    = FALSE;
   weight       = FALSE;
   nbrhd        = FALSE;
   gradient     = FALSE;
   distance_map = FALSE;
   apply_mask   = FALSE;
}
```
In addition, realize that it can be set separately for each verification task. So you if want to disable NetCDF pairs output for one of the obs.field array entries, just include this inside the obs.field array entry:
```
nc_pairs_flag = FALSE;
```

**Comment by jwraby** [comment]
CLASSIFICATION: UNCLASSIFIED

 

John – 

 

Thanks for all the diagnostic work to figure out what is causing the slow performance of Series-Analysis when using the –both option with the nc_pairs NetCDF input and tips on how to get Series-Analysis to run faster through judicious planning of the Grid-Stat runs aimed at producing only the desired data variables in Series-Analysis resulting in smaller NetCDF files. I see the possibilities for configuring Grid-Stat in a way which will better control the fields generated in the NetCDF file from your example below. I will keep this possibility in mind. The study we did, from which this data came,  involved running Grid-Stat for nearly 25 times the input data you saw in my Series-Analysis log file. Instead of just 4 days (24 fcst and obs files per day), out study was for 99 days which involved a lot of Grid-Stat runs to complete on the HPC. So we wanted to be sure we generated everything we might possibly use in those runs. The nc_pairs NetCDF file I provided certainly shows how having all those data variables can add up based on your diagnosis work. Based on our testing of Series-Analysis using the separate –fcst and –obs  inputs, the tradeoff involved in having to generate two separate lists of the input data compared to the value-added of only having to produce one list for the –both option for us would favor going the route of using the –fcst and –obs approach since Series-Analysis executes in about 3 mins using the same amount on input, by taking advantage of the faster processing of grib files as you demonstrated. 

 

I’ve learned a lot from this discussion and I really appreciate the work you and Christina did to address this issue.

 

R/

John

**Reply by JohnHalleyGotway** [reply]
@jwraby, I ran one more test this morning... stripping the FCST and OBS GRIB files down to a single record... and stripping the "both" NetCDF file down to only 2 variables (one for fcst and one for obs). I ran them through Series-Analysis, passing the same input files 1000 times. The total runtime was virtually the same... at 14.8 vs 14.7 seconds. So there's nothing inherently slower in one versus the other.

However, as you've found, there's a significant difference in MET dealing with *extra* unused GRIB records and unused NetCDF variables! I definitely think that we should profile MET NetCDF library code with an eye toward efficiency. I suspect it's reading a lot more information from the file than it actually needs to.

If you have a large dataset of NetCDF matched pairs output file, but running series_analysis is prohibitively slow, you could consider running nccopy as a pre-processing step to strip out only the variable of interest. That obviously isn't a great solution, but it'd allow the series_analysis runs to actually complete.

Thanks,
John

**Comment by jwraby** [comment]
CLASSIFICATION: UNCLASSIFIED

 

John – 

 

Interesting test you conducted. I’m trying to reconcile your results obtained by stripping the FCST and OBS GRIB files down to a single record and running Series-Analysis 1000 times and the same using the –both option with stripped NetCDF input and my results when I use the same inputs with the original (unstripped) grib files as separate fcst and obs inputs. My testing gave a run duration of about 3 mins which is certainly reasonable compare to going the –both route using the NetCDF input files which slows the processing down to a crawl to complete in a few hours.

 

I agree that there may be potential savings achieved by looking at efficiency in the NetCDF library code. I have a large amount of NetCDF output which is populated with a lot of data variables which would have to be dealt with as you suggest. But, I can set up the Series-Analysis run using a config file to generate the same data variables on the fly and using the separate –fcst and –obs grib inputs with the expectation of a realistic run duration. This seems like the easier approach for me right now.

 

R/

John

**Reply by JohnHalleyGotway** [reply]
OK, I wrote up this GitHub issue describing the long delay in reading NetCDF files with lots of variables:
https://github.com/dtcenter/MET/issues/1975

Thanks for letting us know about this unexpected behavior. I'll go ahead and pick one of the previous responses as being a good "Answer" to this discussion and then lock it to prevent future posts. That's how we encourage users to create new discussions for new questions.
